src/js/letters/containers/AddressSection.jsx

```diff
@@ -43,38 +63,122 @@ export class AddressSection extends React.Component {
+        return errorMessage;
+      }
+    }
+
+    // All validations passed; there are no error messages to report
+    return undefined;
+  }
+
+  /**
+   * Runs validation for all fields, returning a complete errorMessages object.
+   *
+   * @return {Object}  Holds all the error messages for all the fields that have them.
+   */
+  validateAll = (address = this.state.editableAddress) => {
+    const errorMessages = {};
+    Object.keys(fieldValidations).forEach((fieldName) => {
+      errorMessages[fieldName] = this.validateField(fieldName, address);
+    });
+
+    return errorMessages;
+  }
+
+  // TODO: Make sure this doesn't allow us to save the address if there are validation errors
+  saveAddress = () => {
+    const errorMessages = this.validateAll();
+    // If there are errors, show them, but don't stop editing and don't save
+    if (Object.keys(errorMessages).length === 0) {
+      this.setState({ errorMessages });
+      return;
+    }
 
+    this.setState({
+      isEditingAddress: false,
+      // Reset all the error messages in case they go to edit again; should be pointless
+      errorMessages: {}
+    });
     this.props.saveAddress(this.state.editableAddress);
   }
 
-  handleChange = (path, update) => {
-    this.setState(({ editableAddress }) => {
-      // reset state code when user updates country
-      const newFragment = (path === 'country'
-        ? {
-          [path]: update,
-          state: '',
-          militaryStateCode: '',
-        }
-        : { [path]: update }
-      );
+  /**
+   * Infers the address type from the address supplied and returns the address
+   *  with the "new" type.
+   */
+  inferAddressType = (address) => {
+    let type = ADDRESS_TYPES.domestic;
+    if (!['USA', 'US'].includes(address.country)) {
+      type = ADDRESS_TYPES.international;
+    } else if (address.militaryStateCode) {
+      // TODO: Make sure we clear this out if a state code is selected
+      type = ADDRESS_TYPES.military;
+    }
 
-      return {
-        editableAddress: Object.assign({}, editableAddress, newFragment),
-      };
+    return Object.assign({}, address, { type });
+  }
+
+  handleChange = (fieldName, update) => {
+    // When a field is changed, make sure we validate it
+    if (!this.state.shouldValidate[fieldName]) {
+      this.setState({ shouldValidate: Object.assign({}, this.state.shouldValidate, { [fieldName]: true }) });
+    }
+
+    let address = Object.assign({}, this.state.editableAddress, { [fieldName]: update });
+    // if country is changing we should clear the state
+    if (fieldName === 'country') {
+      address.state = '';
+    }
+
+    // Make sure we've got the right address type (domestic, military, international)
+    address = this.inferAddressType(address);
+
+    // Update the error messages
+    // TODO: This might get super slow, so we can debounce this part if necessary...probably
+    const errorMessages = this.validateAll(address);
```

Comment:
Is there a reason we can't run `this.validateField(fieldName, address)` here instead of running the validations on all fields?

Reply:
Yeah, I'd prefer to do that, but the reason I have it this way is because if one field changes, it may change the validations on other fields.

For example, we might have a validation that says the city needs to be a valid `militaryPostOfficeTypeCode` (FPO and friends, I think) if it's a military address. So if we set our country to USA and state to AE, but city to San Francisco, there'll be a validation error on the city, but as soon as we change the state to California, we want that validation error to go away, so if we don't validate all the fields, we'd need to touch the city again to reset the validation.

To keep it as performant as possible, I keep track of all the fields that have been modified so we only run validation on them instead of _all_ the fields. I _think_ that'll be okay, but...I haven't really tested it out extensively, if I'm honest. 😳

Reply:
Gotcha, that makes sense! We can test more today during QA testing.